core/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -1524,19 +1538,62 @@ public void waitForOpsToComplete(final long seqNo) throws InterruptedException {
     /**
      * Marks the shard with the provided allocation ID as in-sync with the primary shard. See
      * {@link org.elasticsearch.index.seqno.GlobalCheckpointTracker#markAllocationIdAsInSync(String, long)}
-     * for additional details.
+     * for additional details. Because this operation could be completed asynchronously on another thread, the caller must provide a latch;
+     * this latch will be counted down after the operation completes.
      *
-     * @param allocationId    the allocation ID of the shard to mark as in-sync
-     * @param localCheckpoint the current local checkpoint on the shard
+     * @param allocationId      the allocation ID of the shard to mark as in-sync
+     * @param localCheckpoint   the current local checkpoint on the shard
+     * @param latch             a latch that is counted down after the operation completes
+     * @param onFailureConsumer a callback if an exception occurs completing the operation
      */
-    public void markAllocationIdAsInSync(final String allocationId, final long localCheckpoint) throws InterruptedException {
-        verifyPrimary();
-        getEngine().seqNoService().markAllocationIdAsInSync(allocationId, localCheckpoint);
+    public void markAllocationIdAsInSync(
```

Comment:
I'm not a fan of this api change. All other api does acquire a permit inline. How about using this code in `RecoverySourceHandler#finalizeRecovery`: 

```
PlainActionFuture<Releasable> permitFuture = new PlainActionFuture<>();
shard.acquirePrimaryOperationPermit(permitFuture, ThreadPool.Names.SAME);
try (Releasable permit = permitFuture.actionGet()) {
    if (shard.state() == IndexShardState.RELOCATED) {
        throw new DelayRecoveryException();
    }
    shard.markAllocationIdAsInSync(request.targetAllocationId(), targetLocalCheckpoint);
}
```